### PR TITLE
Add CodeFix providers for warnings BC40003 and CS0108

### PR DESCRIPTION
**Pull request by Giftednewt**
Added CodeFixProvider to fix issue #3437 and also added a CodeFixProvider to provide a fix for CS0108

### CODE REVIEW

File: src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        internal const string CS0108 = "CS0108"; // 'SomeClass.SomeMember' hides inherited member 'SomeClass.SomeMember'. Use the new keyword if hiding was intended.
+
+        public override ImmutableArray<string> FixableDiagnosticIds
+        {
+            get
+            {
+                return ImmutableArray.Create(CS0108);
+            }
+        }
```
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.cs]: can just be `public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CS0108);`

File: src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                newNode = fieldDeclaration.AddModifiers(SyntaxFactory.Token(SyntaxTriviaList.Empty, SyntaxKind.NewKeyword, SyntaxTriviaList.Create(SyntaxFactory.Whitespace(" "))));
+            }
+
+            var cleanupService = document.GetLanguageService<ICodeCleanerService>();
+
+            if (cleanupService != null && newNode != null)
+            {
+                newNode = cleanupService.Cleanup(newNode, new[] { newNode.Span }, document.Project.Solution.Workspace, cleanupService.GetDefaultProviders(), cancellationToken);
+            }
```
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.cs]: It is not necessary to run the ICodeCleanerService.  You can just add the [Formatter.Annotation](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Formatting/Formatter.cs,24) to the node you return, and the node will be formatted with the users formatting options automatically when the formatter runs.
example:
`return newNode.WithAdditionalAnnotations(Formatter.Annotation);`

File: src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.AddNewKeywordAction.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                _newNode = newNode;
+            }
+
+            protected override async Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
+            {
+                var root = await _document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var newRoot = root.ReplaceNode(_node, _newNode);
+                var newDocument = await Formatter.FormatAsync(_document.WithSyntaxRoot(newRoot)).ConfigureAwait(false);
+
```
Comment (jmarolf) [inline src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.AddNewKeywordAction.cs]: If you add the Formatter Annotation then you don't need to format the document yourself.  However, you probably want to add the trivia from the old node.
`var newRoot = root.ReplaceNode(_node, _newNode.WithTriviaFrom(_node));`
That way if the user had a comment in front or behind this method/property/field it is preserved after the replacement is done.

File: src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.vb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            Dim methodStatement = TryCast(node, MethodStatementSyntax)
+            If methodStatement IsNot Nothing Then
+                newNode = methodStatement.AddModifiers(SyntaxFactory.Token(SyntaxKind.OverloadsKeyword))
+            End If
+
+            Dim cleanupService = document.GetLanguageService(Of ICodeCleanerService)
+
+            If cleanupService IsNot Nothing AndAlso newNode IsNot Nothing Then
+                newNode = cleanupService.Cleanup(newNode, {newNode.Span}, document.Project.Solution.Workspace, cleanupService.GetDefaultProviders(), cancellationToken)
+            End If
```
Comment (jmarolf) [inline src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.vb]: Same comments as the C# side, you shouldn't need to run cleanup on this.

File: src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.AddOverloads.vb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                _node = node
+                _newNode = newNode
+            End Sub
+
+            Protected Overrides Async Function GetChangedDocumentAsync(cancellationToken As CancellationToken) As Task(Of Document)
+                Dim root = Await _document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
+
+                Dim newRoot = root.ReplaceNode(_node, _newNode)
+                Dim newDocument = Await Formatter.FormatAsync(_document.WithSyntaxRoot(newRoot), cancellationToken:=cancellationToken).ConfigureAwait(False)
+
```
Comment (jmarolf) [inline src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.AddOverloads.vb]: Same comment as on the C# side, you don't want to run the formatter yourself, and you should try and preserve the users trivia for them.

File: src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.vb (hunk 4 above)
Comment (Giftednewt) [inline src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.vb]: The reason that I was running cleanup is that when the Overloads modifier is added to the token it the modifiers are not ordered in the way that the cleanup does internally therefore after the code fix is applied the formatter then re-orders the modifiers. This is as per the discussion in #3437

Comment (jmarolf) [inline src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.vb]: I see, that makes sense.  Could you add a comment then explaining this?

File: src/Features/CSharp/Portable/CSharpFeaturesResources.resx [hunk 6]
```diff
@@ -437,4 +437,7 @@
   <data name="ERR_NameNotInContext" xml:space="preserve">
     <value>The name '{0}' does not exist in the current context.</value>
   </data>
+  <data name="HideBase" xml:space="preserve">
+    <value>Hide base member</value>
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CSharpFeaturesResources.resx]: @DustinCampbell What do you think of this as the Code Action string?

File: src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.AddNewKeywordAction.cs (hunk 3 above)
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.AddNewKeywordAction.cs]: Right - please don't format the whole document, as this can cause unrelated changes elsewhere in the file to show up as part of your fix.

File: src/Features/CSharp/Portable/CSharpFeaturesResources.resx (hunk 6 above)
Comment (DustinCampbell) [inline src/Features/CSharp/Portable/CSharpFeaturesResources.resx]: Hmmm... "Hide" or "Shadow"? I think I'm OK with either, so Hide is fine with me. What do you think?

File: src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                originalNode = token.GetAncestor<FieldDeclarationSyntax>();
+            }
+
+            if(originalNode == null)
+            {
+                return;
+            }
+
+            var newNode = GetNewNode(context.Document, originalNode, context.CancellationToken);
```
Comment (Pilchie) [inline src/Features/CSharp/Portable/CodeFixes/HideBase/HideBaseCodeFixProvider.cs]: Ideally, you would do as much of the work as possible *inside* `AddNewKeywordAction`, so that it isn't executed when determining whether or not to show the lightbulb, and instead, only when `GetChangedDocument` is called to create the preview.

File: src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.vb (hunk 4 above)
Comment (Pilchie) [inline src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.vb]: Also, as above, ideally you would move this into the `GetChangedDocument` so that it doesn't run as part of determining lightbulb visibility.

File: src/Features/CSharp/Portable/CSharpFeaturesResources.resx (hunk 6 above)
Comment (Pilchie) [inline src/Features/CSharp/Portable/CSharpFeaturesResources.resx]: Spec section 10.3.4 calls it hiding, so let's go with that.

---
A class-member-declaration is permitted to declare a member with the same name or signature as an inherited member. When this occurs, the derived class member is *said to hide* the base class member. Hiding an inherited member is not considered an error, but it does cause the compiler to issue a warning. To suppress the warning, the declaration of the derived class member can include a new modifier to indicate that the derived member is intended to hide the base member. This topic is discussed further in §3.7.1.2.
If a new modifier is included in a declaration that doesn’t hide an inherited member, a warning to that effect is issued. This warning is suppressed by removing the new modifier.

**Comment by DustinCampbell** [comment]
test mac please

**Comment by Giftednewt** [comment]
I'd originally put the whitespace trivia in because the result displayed in the preview window wasn't formatted correctly. I have tested it again now without the whitespace and it's formatting as expected. I'll get that removed and update the pr.

Looks like the mac test failed due to the following files.
inflating: packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/src/XMakeTasks/CodeTaskFactory.cs
packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/src/XMakeTasks/UnitTests/XamlTaskFactory_Tests.cs

The test result states the following
Nondeterministic failure occasionally seen building coreclr. Rerunning should fix the issue

Am I able to manually re-run the tests?

**Comment by Pilchie** [comment]
The actual mac failure seems to be:
```
/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/Microsoft.CSharp.CurrentVersion.targets(383,9): error MSB6003: The specified task executable "csc2.exe" could not be run. ApplicationName='/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/tools/../tools/csc2.exe', CommandLine=' -noconfig @"/var/folders/y3/7p46jgrx2nv63czf30tf_t1w0000gn/T/tmp23f03e748bb34f249eea61ac453e8572.rsp"', CurrentDirectory='', Native error= Cannot find the specified file [/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/src/Compilers/Core/Portable/CodeAnalysis.csproj]
/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/Microsoft.CSharp.CurrentVersion.targets(383,9): error MSB6003: The specified task executable "csc2.exe" could not be run. ApplicationName='/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/tools/../tools/csc2.exe', CommandLine=' -noconfig @"/var/folders/y3/7p46jgrx2nv63czf30tf_t1w0000gn/T/tmp9dbb62264ed44847a3e1187e05643a2e.rsp"', CurrentDirectory='', Native error= Cannot find the specified file [/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj]
/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/Microsoft.CSharp.CurrentVersion.targets(383,9): error MSB6003: The specified task executable "csc2.exe" could not be run. ApplicationName='/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/tools/../tools/csc2.exe', CommandLine=' -noconfig @"/var/folders/y3/7p46jgrx2nv63czf30tf_t1w0000gn/T/tmp59cb5df4d355497181fc9e7b068bb402.rsp"', CurrentDirectory='', Native error= Cannot find the specified file [/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/src/Tools/Source/CompilerGeneratorTools/Source/BoundTreeGenerator/CompilersBoundTreeGenerator.csproj]
/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/Microsoft.CSharp.CurrentVersion.targets(383,9): error MSB6003: The specified task executable "csc2.exe" could not be run. ApplicationName='/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/packages/Microsoft.Build.Mono.Debug.14.1.0.0-prerelease/lib/tools/../tools/csc2.exe', CommandLine=' -noconfig @"/var/folders/y3/7p46jgrx2nv63czf30tf_t1w0000gn/T/tmp572d5a23cb03480ab304cc30e53f21ea.rsp"', CurrentDirectory='', Native error= Cannot find the specified file [/Users/dotnet-bot/j/workspace/dotnet_roslyn_prtest_mac/src/Tools/Source/CompilerGeneratorTools/Source/CSharpErrorFactsGenerator/CSharpErrorFactsGenerator.csproj]
```
@agocke does this look familiar?

File: src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.AddOverloads.vb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                Dim methodStatement = TryCast(node, MethodStatementSyntax)
+                If methodStatement IsNot Nothing Then
+                    newNode = methodStatement.AddModifiers(SyntaxFactory.Token(SyntaxKind.OverloadsKeyword))
+                End If
+
+                'Make sure we preserve any trivia from the original node
+                newNode = newNode.WithTriviaFrom(node)
+
+                'We need to perform a cleanup on the node because AddModifiers doesn't adhere to the VB modifier ordering rules
+                Dim cleanupService = document.GetLanguageService(Of ICodeCleanerService)
```
Comment (jmarolf) [inline src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.AddOverloads.vb]: Is there a reason this is before the CodeCleanerService runs?

**Comment by agocke** [comment]
@Pilchie Sorry, never seen it before -- it looks like some machine configuration issue. I can't imagine this change would have caused that.

**Comment by agocke** [comment]
Based on the error message, it looks like csc2.exe is missing from the NuGet restore?

**Comment by Pilchie** [comment]
test mac please

File: src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.AddOverloads.vb (hunk 8 above)
Comment (Giftednewt) [inline src/Features/VisualBasic/Portable/CodeFixes/OverloadBase/OverloadBaseCodeFixProvider.AddOverloads.vb]: My intention was that if the original trivia needed formatting/cleaning up that it would be done with the call to cleanup. I suppose that it could also be done after the CodeCleanerService runs as I'm now adding the Formatter annotation which would may save any unintentional changes to the trivia.

**Comment by jmarolf** [comment]
test all please

**Comment by mmitche** [comment]
Can one of the admins approve this PR test or whitelist this user?

**Comment by Pilchie** [comment]
@Giftednewt (or @jmarolf) can you merge the conflicts and re-push, then we'll see whether we can get Jenkins to like it.

**Comment by Giftednewt** [comment]
@Pilchie I'll get that merged up when I get back home in about an hour or so

**Comment by jmarolf** [comment]
:+1: @Pilchie any objections to merging this in?

**Comment by Pilchie** [comment]
@dotnet-bot okay to test

**Comment by jmarolf** [comment]
@dotnet-bot test vsi please

**Comment by Pilchie** [comment]
No!  Let's do it!
